packages/buidler/contracts/CLR.sol

```diff
@@ -0,0 +1,162 @@
+        require(roundStart == 0, "Round already opened");
+        _;
+    }
+
+    modifier isRoundOpen() {
+        require(
+            getBlockTimestamp() < roundStart.add(roundDuration),
+            "Round is not open"
+        );
+        _;
+    }
+
+    modifier isRoundClosed() {
+        require(
+            roundStart != 0 &&
+                getBlockTimestamp() >= roundStart.add(roundDuration),
+            "Round is not closed"
+        );
+        _;
+    }
+
+    function startRound(uint256 _roundDuration) public onlyOwner {
+        roundStart = getBlockTimestamp();
+        roundDuration = _roundDuration;
+        emit RoundStarted(roundStart, roundDuration);
+    }
+
+    function getBlockTimestamp() public view returns (uint256) {
+        return block.timestamp;
+    }
+
+    function addRecipient(address payable addr, string memory data)
+        public
+        onlyOwner
+        beforeRoundOpen
+        returns (uint256)
+    {
+        Recipient memory r = Recipient({
+            addr: addr,
+            data: data,
+            totalDonation: 0,
+            sumOfSqrtDonation: 0,
+            matchingWeight: 0,
+            withdrawn: false
+        });
+        recipients.push(r);
+
+        uint256 index = recipients.length - 1;
+        emit RecipientAdded(addr, data, index);
+        return index;
+    }
+
+    function allowDonor(address donor) public onlyOwner {
+        donorAllowList[donor] = true;
+    }
+
+    function blockDonor(address donor) public onlyOwner {
+        delete donorAllowList[donor];
+    }
+
+    function donate(uint256 index) public payable isRoundOpen {
+        require(donorAllowList[msg.sender], "CLR:donate not in donorAllowList");
+
+        recipients[index].totalDonation = recipients[index].totalDonation.add(
+            msg.value
+        );
+        recipients[index].sumOfSqrtDonation = recipients[index]
+            .sumOfSqrtDonation
+            .add(Math.sqrt(msg.value));
+    }
+
+    // this function could be run in multiple steps paying attention to msg.gas (remaining gas)
+    function calculateMatching(uint256 toIndex) public isRoundClosed {
+        require(
+            calculatedToIndex < recipients.length,
+            "CLR:calculate already finished calculating"
+        );
+
+        if (toIndex > recipients.length) {
+            toIndex = recipients.length;
+        }
+        for (uint256 i = calculatedToIndex; i < toIndex; i++) {
+            recipients[i].matchingWeight = recipients[i].sumOfSqrtDonation.mul(
+                recipients[i].sumOfSqrtDonation
+            );
+            totalMatchingWeight = totalMatchingWeight.add(
+                recipients[i].matchingWeight
+            );
+        }
+
+        calculatedToIndex = toIndex;
+        if (calculatedToIndex >= recipients.length) {
+            emit MatchingCalculated();
+        }
+    }
+
+    function withdraw(uint16 index) public isRoundClosed returns (uint256) {
```

Comment:
rename this withdraw to "recipientWithdraw". Have a new withdraw for admin to withdraw what's left.